readers/api/endpoint.go

```diff
@@ -80,3 +110,118 @@ func restoreEndpoint(svc readers.MessageRepository) endpoint.Endpoint {
 		return restoreMessagesRes{}, nil
 	}
 }
+
+func generateCSV(page readers.MessagesPage) ([]byte, error) {
+	var buf bytes.Buffer
+	writer := csv.NewWriter(&buf)
+
+	header := []string{
+		"Channel",
+		"Subtopic",
+		"Publisher",
+		"Protocol",
+		"Name",
+		"Unit",
+		"Value",
+		"String_value",
+		"Bool_value",
+		"Data_value",
+		"Sum",
+		"Time",
+		"Update_time",
+	}
+	if err := writer.Write(header); err != nil {
+		return nil, err
+	}
+
+	for _, msg := range page.Messages {
+		switch m := msg.(type) {
+		// handle SenML messages
+		case senml.Message:
+			row := []string{
+				m.Channel,
+				m.Subtopic,
+				m.Publisher,
+				m.Protocol,
+				m.Name,
+				m.Unit,
+				getFloatValue(m.Value, ""),
+				getStringValue(m.StringValue, ""),
+				getBoolValue(m.BoolValue, ""),
+				getStringValue(m.DataValue, ""),
+				getFloatValue(m.Sum, ""),
+				fmt.Sprintf("%v", m.Time),
+				fmt.Sprintf("%v", m.UpdateTime),
+			}
+			if err := writer.Write(row); err != nil {
+				return nil, err
+			}
+		// handle JSON messages
+		case map[string]interface{}:
+			row := make([]string, len(header))
+			for key, value := range m {
+				if idx := indexOf(header, key); idx != -1 {
+					row[idx] = getValue(value, "")
+				}
+			}
+			if err := writer.Write(row); err != nil {
+				return nil, err
+			}
+		default:
+			continue
+		}
+	}
+
+	writer.Flush()
+	if err := writer.Error(); err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// Helper function to handle the different types of values.
+func getValue(value interface{}, defaultValue string) string {
+	switch v := value.(type) {
+	case *float64:
+		return getFloatValue(v, defaultValue)
+	case *string:
+		return getStringValue(v, defaultValue)
+	case *bool:
+		return getBoolValue(v, defaultValue)
+	default:
+		return fmt.Sprintf("%v", v)
+	}
+}
+
+// Helper function to get the index of a string in a slice.
+// Returns -1 if the string is not found in the slice.
+func indexOf(slice []string, item string) int {
+	for i, val := range slice {
+		if val == item {
+			return i
+		}
+	}
+	return -1
+}
+
+func getStringValue(ptr *string, defaultValue string) string {
+	if ptr != nil {
+		return *ptr
+	}
+	return defaultValue
+}
+
+func getFloatValue(ptr *float64, defaultValue string) string {
+	if ptr != nil {
+		return fmt.Sprintf("%v", *ptr)
+	}
+	return defaultValue
+}
+
+func getBoolValue(ptr *bool, defaultValue string) string {
+	if ptr != nil {
+		return fmt.Sprintf("%v", *ptr)
+	}
+	return defaultValue
+}
```

Comment:
This can be probably done with a single method:
```
func valueToString(ptr *bool, defaultValue string) string {
	if ptr != nil {
		return fmt.Sprintf("%v", *ptr)
	}
	return defaultValue
}
```